site/en/articles/renderingng-fragmentation/index.md

```diff
@@ -0,0 +1,204 @@
+
+We shipped the first part of LayoutNG in 2019, which consisted of regular block container layout, inline layout, floats and out-of-flow positioning, but no support for flex, grid, or tables, and no block fragmentation support at all. We would fall back to using the legacy layout engine for flex, grid, tables, plus anything that involved block fragmentation. That was true even for block, inline, floating and out-of-flow elements within fragmented content—as you can see, upgrading such a complex layout engine in-place is a very delicate dance.
+
+In addition, believe it or not, by mid-2019 the majority of the core functionality of LayoutNG block fragmentation layout was already implemented (behind a flag). So, why did it take so long for it to ship? The short answer is: fragmentation has to coexist correctly with various legacy parts of the system, which cannot be removed or upgraded until all dependencies are upgraded. For the long answer, see the following details.
+
+### Legacy engine interaction
+
+Legacy data structures are still in charge of JavaScript APIs that read layout information, so we need to write back data to the legacy engine in a manner that it understands. This includes updating the legacy multi-column data structures, like [LayoutMultiColumnFlowThread](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/layout/layout_multi_column_flow_thread.h;l=144;drc=237a52e7e2b6d22eb1b5912e134384254d5b8974), correctly. 
+
+### Legacy engine fallback detection and handling
+
+We had to fall back to the legacy layout engine when there was content inside that couldn't yet be handled by LayoutNG block fragmentation. At the time of shipping core LayoutNG block fragmentation (spring 2022), that included flex, grid, tables, and anything that's printed. This was particularly tricky because we need to detect the need for legacy fallback before creating [objects in the layout tree](/articles/layoutng/#a-30000-foot-view-of-layout-engine-architectures). For example, before we know if there's a multi-column container ancestor, and before we know which DOM nodes will become a [formatting context](https://developer.mozilla.org/docs/Web/Guide/CSS/Block_formatting_context) or not. It's a chicken-and-egg problem that doesn't have a perfect solution, but as long as its only misbehavior is false positives (fallback to legacy when there's actually no need), it's ok, because any bugs in that layout behavior are ones Chromium already has, not new ones.
```

Comment:
```suggestion
We had to fall back to the legacy layout engine when there was content inside that couldn't yet be handled by LayoutNG block fragmentation. At the time of shipping core LayoutNG block fragmentation (spring 2022), that included flex, grid, tables, and anything that's printed. This was particularly tricky because we needed to detect the need for legacy fallback before creating [objects in the layout tree](/articles/layoutng/#a-30000-foot-view-of-layout-engine-architectures). For example, we needed to detect before we knew if there was a multi-column container ancestor, and before we knew which DOM nodes would become a [formatting context](https://developer.mozilla.org/docs/Web/Guide/CSS/Block_formatting_context) or not. It's a chicken-and-egg problem that doesn't have a perfect solution, but as long as its only misbehavior is false positives (fallback to legacy when there's actually no need), it's ok, because any bugs in that layout behavior are ones Chromium already has, not new ones.
```